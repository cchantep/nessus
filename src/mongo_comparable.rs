// MongoComparable provides evidence that a value of type A can be compared with a value of type B
// in MongoDB queries. This trait is used at the struct level to indicate that specific fields
// can be compared with specific types of values in MongoDB operations.
pub trait MongoComparable<A, B> {}
