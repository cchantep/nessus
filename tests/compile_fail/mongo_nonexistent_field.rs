// This test verifies that MongoComparable can't be used with a non-existent field

use nessus::{FieldWitnesses, MongoComparable};
use nessus::mongo_comparable::MongoComparable as MongoComparableTrait;
use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize, FieldWitnesses, MongoComparable)]
struct Product {
    Name: String,
    Price: f64,
}

// Define a field witness for a nonexistent field
#[allow(non_camel_case_types)]
struct Weight;

impl nessus::field_witnesses::FieldName for Weight {
    fn field_name() -> &'static str { "Weight" }
}

fn main() {
    // This should fail because Product doesn't have a Weight field
    // The MongoComparable trait requires HasField to be implemented
    fn assert_implements_mongo_comparable<T, A, B>()
    where
        T: MongoComparableTrait<A, B>
    {}
    
    assert_implements_mongo_comparable::<Product, 
        <Product as nessus::field_witnesses::HasField<Weight>>::Value, 
        f64>();
}
