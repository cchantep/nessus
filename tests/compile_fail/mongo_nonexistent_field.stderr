error[E0277]: the trait bound `Product: HasField<Weight>` is not satisfied
  --> tests/compile_fail/mongo_nonexistent_field.rs:30:10
   |
30 |         <Product as nessus::field_witnesses::HasField<Weight>>::Value,
   |          ^^^^^^^ the trait `HasField<Weight>` is not implemented for `Product`
   |
   = help: the following other types implement trait `HasField<F>`:
             `Product` implements `HasField<Name>`
             `Product` implements `HasField<Price>`
